flocker/control/_model.py

```diff
@@ -504,6 +504,60 @@ def get_node(deployment, uuid, **defaults):
     return get_node
 
 
+class Lease(PClass):
+    """
+    A lease.
+    """
+    dataset_id = field(type=UUID)
+    node_uuid = field(type=UUID)
+    expiration = field(type=datetime, optional=True)
+
+
+class Leases(PMap):
+    """
+    Leases.
+
+    Map dataset IDs to Leases.
+    """
+    __key_type__ = UUID
+    __value_type__ = Lease
+    __invariant__ = _keys_match_dataset_id
+
+    def _check_lease(self, dataset_id, node_uuid):
+        if dataset_id in self and self[dataset_id].node_uuid != node_uuid:
+            raise ValueError("Lease already held by another node")
```

Comment:
It's important that this error say which node, so the lock can be broken if necessary.